Parse review comments to handle /ok-to-test

matthyx opened the pull request:
Fixes #3827 
For the moment review.go and review_test.go are almost copy/paste from ic.go and ic_test.go
I plan to refactor ic* pr* and review* in a later PR to factor out redundant code.

matthyx commented:
/cc @cjwagner

matthyx commented on prow/github/fakegithub/fakegithub.go:
```diff
@@ -235,7 +235,7 @@ func (f *FakeClient) GetIssueLabels(owner, repo string, number int) ([]github.La
 	re := regexp.MustCompile(fmt.Sprintf(`^%s/%s#%d:(.*)$`, owner, repo, number))
 	la := []github.Label{}
-	for _, l := range f.LabelsAdded {
+	for _, l := range append(f.LabelsAdded, f.ExistingLabels...) {
```
Oh, you're right... I thought append() was a pure function by the look of it.
Bummer, I will probably end up with something inelegant here.

matthyx commented on prow/plugins/sigmention/sigmention_test.go:
```diff
@@ -124,7 +124,7 @@ func TestSigMention(t *testing.T) {
 			body:              "Code Comment.  Design Review\n@kubernetes/sig-node-proposals\ncc @kubernetes/sig-api-machinery-bugs",
 			expectedRepeats:   []string{"@kubernetes/sig-node-proposals", "@kubernetes/sig-api-machinery-bugs"},
-			expectedNewLabels: []string{"sig/node", "sig/api-machinery", labels.Bug},
+			expectedNewLabels: []string{"kind/design", "sig/node", "sig/api-machinery", labels.Bug},
```
For strange reason it got added, I will investigate more -  but it is maybe a side effect of the append before.

matthyx replied:
Alright, this comes from the name of the SIG `sig-node-proposals` that matches the `kindMap` entry "proposals" resulting in label `kind/design` being added at line 128.
Before the change in fakegithub.AddLabel() the label wasn't added because we have f.ExistingLabels != nil and the test only allows to add label to the issue if it already exists on the repo.
I am actually wondering what should be the correct behaviour here?

cjwagner replied:
Ah, I was confused about this. I thought `ExistingLabels` were the labels that were preexisting on an issue before the test, but they are actually the set of labels that exist in the repo.  So the change you made to `fakegithub` isn't correct.

matthyx replied:
Ok, added a `LabelsExisting` array for that purpose (sig) and will make a PR to rename `ExistingLabels` to avoid this confusion.

matthyx replied:
#10062

matthyx commented:
@cjwagner so as understant it, I could replace completely `ic.go` with this file (probably renamed `generic.go`) and handle all comments (issue, pr and review) in one function?

cjwagner commented:
> @cjwagner so as understant it, I could replace completely ic.go with this file (probably renamed generic.go) and handle all comments (issue, pr and review) in one function?

Exactly. However, keep in mind that GenericCommentEvents also include comments in PR body text in addition to the issue/pr/review comments.

matthyx commented:
Alright, I have just done the few nits suggested, here is what remains:
- get rid of IssueCommentEvent handling (all should be processed by GenericCommentEvent)
- clarify the behaviour of AddLabels() in fakegithub
- check if sigmention should really try to add the label `kind/design`

I will be off 3 days for a conference, maybe @cjwagner you could comment on the last 2 points?

fejta commented:
/uncc
Cole/Matt add me back if you want me to review

matthyx commented:
> Exactly. However, keep in mind that GenericCommentEvents also include comments in PR body text in addition to the issue/pr/review comments.

Is that enough to skip generic events that have `Body` equals `IssueBody`, corresponding to `handlePullRequestEvent` and `handleIssueEvent`?

cjwagner commented:
> @cjwagner is that enough to skip generic events that have Body equals IssueBody, corresponding to handlePullRequestEvent and handleIssueEvent?

Yeah, I think that will work. It seems kind of hacky, and there could be some weird edge cases if the PR body is edited, but that could be ok.
Keep in mind that handling the event twice shouldn't produce any weird behavior in this case, it just wastes some API tokens and only in a rare case. So if the hacky solution will ever produce strange behavior it might be better to just waste some API tokens and handle the event twice.
I don't feel strongly either way on this so I'll leave it to your judgement.

matthyx commented:
> I don't feel strongly either way on this so I'll leave it to your judgement.

Alright so let's play it safe... I have removed the check `Body == IssueBody` and added a second run of `handleGenericComment()` in the tests to detect inconsistencies.
It seems to work, but required some adjustments in `fakegithub.go` as you will see.

cjwagner commented on prow/plugins/trigger/generic-comment_test.go:
```diff
@@ -612,34 +617,30 @@ func TestHandleIssueComment(t *testing.T) {
+		if err := handleGenericComment(c, nil, event); err != nil {
+			t.Fatalf("Didn't expect error: %s", err)
+		}
+		if err := handleGenericComment(c, nil, event); err != nil {
```
If we run the handler twice for every test case we may cover up errors that occur if the handler is only run once (the normal case). The easiest way to catch this kind of bug in the unit tests would be to move all the validation checks into a local function and then use the function to validate state after both handler calls **and** between the first and second handler call.

matthyx replied:
Indeed, and it has uncovered a small bug!
